docs/v1.md

```diff
@@ -1,7 +1,8 @@
-# AsyncAPI Parser API v1.0.0-alpha.3
+# AsyncAPI Parser API v1.0.0-alpha.4
 
 ## AsyncAPIDocument
 - channels() : `Channels`
+- components() : `Components`
```

Comment:
I'm not 100% sure this is a good idea on an intent-driven API. If we forget about AsyncAPI, what does `components` mean? Same probably goes for `extensions` 🤔

Reply:
That's a really good point. And related to the discussion in https://github.com/asyncapi/parser-js/pull/612/files#r971793485. 
`components()` use case is only internal, for the schemas iterator on `parser-js` (See [code here](https://github.com/asyncapi/parser-js/blob/next-major/src/iterator.ts#L62)).

I think we deviated a bit from the Intent-Driven approach with this and went direct to mimic current `parser-js` behavior just to provide same functionalities (assigning anonymous names when missing). 
I agree `components()` doesn't represent any intent and there are no more use cases found than the one mentioned above. 

If we apply what you @fmvilas suggested in https://github.com/asyncapi/parser-api/pull/71/files#r972687041, and then we modify the iterator a bit (replacing `components` by those new statuses in the `SchemaTypesToIterate`)  we could keep the functionality of the iterator and get rid of the `components()` method in favor of those filters on each collection.

Regarding `extensions()`, I see it should be kept. How would you grab extensions from the root doc then?

cc @magicmatatjahu

Reply:
Components are needed for the Studio and React Component, not for iterator. Iterator is internal functionality.

Reply:
In Studio we render navigation on the left, atm channels, servers, messages and schemas, these two last point to the `components.messages` and `components.schemas` objects.

Reply:
> these two last point to the `components.messages` and `components.schemas` objects

Or in another way, we are showing Messages and Schemas that do have an ID (all except those declared under the operation).
Even though this doesn't seem ok to me (I think we should also show the ones with anonymous ID, but not the place to discuss), you can still show them by either filter them by their name (remove those with anonymous ID) or similar.

Reply:
From private discussion:

> how you will retrieve correlationsIds and other things from components, without components() ? Will you handle raw JSON from AsyncAPI or write function in AsyncAPiDocument  like we do for schemas , messages  etc? We need components data for FormBased AsyncAPI creator and also for Drag&Drop Builder.

If even from the point of view of templates a `components()` function is not needed, then for Studio and Html-Template it is (and similar tools) is, even for bundler or optimizer (we can use new parser to handle some things for which we have to write custom logic in mentioned packages).

Reply:
The solution for that would be to move all the methods from `Components` to the root doc `AsyncAPI`. Some of them are already there, like `channels()`, but the following not:

- channelParameters(): `ChannelParameters`
- serverVariables(): `ServerVariables`
- operationTraits(): `OperationTraits`
- messageTraits(): `MessageTraits`
- correlationIds(): `CorrelationIds`
- securitySchemes(): `SecuritySchemes`
- serverBindings(): `Map[string, Bindings]`
- channelBindings(): `Map[string, Bindings]`
- operationBindings(): `Map[string, Bindings]`
- messageBindings(): `Map[string, Bindings]`

TBH, I now see it makes sense to move all of those to the root of the document, but let's go deep on this and find the pros and cons. 

PROS:
- Decoupling the API from what the document looks like. This is one of the main purposes of this new API.
- Related to the previous one; no breaking change if we ever remove a component, let's say Operations. Not very probable but could happen.
- If we forget about how a document looks like, it feels more natural (ergo better UX) to call `doc.correlationIds()` than `doc.components().correlationIds()`.
- As a remember, users can still use the same objects that were previously located under `Components` object though those methods.

CONS:
- Use cases where, for some reason, **all components** need to be accessed at once, like printing/showing all components in the Studio nav bar will now need to call all those methods individually where, maybe, previously the call was done by just calling **all** methods within the object (`Components`).
- Sure there are more CONS, so I need your help.

Can you folks help me finding more cons or to add weight enough to the one I mentioned? 

cc @fmvilas @magicmatatjahu @jonaslagoni

Reply:
@smoya I have one problem, how you will handle extensions in `Components` object if components will be removed? `extensions()` in root will return them? In this case methods like `filterInUse` has don't any sense, because extensions are strictly related to given object not to "collections" of objects, so `extensions()` in root should return only extensions from root, not from root + components. For me, removing components does not make sense at all and I would leave this model, but I look from pov of the Studio and the tools/ideas for them that I know of, which components they need.

Reply:
What about a `componentsExtensions()` method? If the only method inside `components` is going to be `extensions()` then just merge them into one.

I like the new structure @smoya proposed. For the downside "...all components need to be accessed at once, like printing/showing all components in the Studio nav bar will now need to call all those methods individually...", I don't think it's really a problem in the case of Studio. Those components inside the sidebar should be individual components and therefore they should only grab the info they need, not the whole components section.

Reply:
Calling all these methods (from components perspective) will be needed anyway, the only significant difference between the Components model and the methods at the level of the AsyncAPIDocument model is that you will need to filter these objects whether they are in components or not - navigation in Postman/Stoplight displays if given object(s) is/are in components or not and you can for example remove it/add new or go to given line in components (and that's I have in mind when it comes to Studio), so if we wanna go with method on root, I opt to having also `filterByInComponents()` method whether they are referenced in document or not.

Reply:
Agreed. But like I said in the comment below, let's not tie things to location (components) but to meaning (optional). So yeah, `filterByOptional()` should give you everything that's in components.